File: backend/app/api/sse.py
from __future__ import annotations
import asyncio
import json
from sse_starlette.sse import EventSourceResponse


async def event_stream(queue: asyncio.Queue):
    while True:
        event = await queue.get()
        yield {"data": json.dumps(event, default=str)}
        if event.get("kind") == "run.complete" or event.get("kind") == "run.error":
            break


def sse_response(queue: asyncio.Queue):
    return EventSourceResponse(event_stream(queue))

File: frontend/src/api/sse.ts
import type { RunEvent } from "../types/events.ts";

export function subscribeToRun(runId: string, onEvent: (e: RunEvent) => void): () => void {
  const es = new EventSource(`/api/runs/${runId}/events`);
  es.onmessage = (msg) => {
    try {
      const data = JSON.parse(msg.data) as RunEvent;
      onEvent(data);
      if (data.kind === "run.complete" || data.kind === "run.error") {
        es.close();
      }
    } catch (e) {
      console.error("bad event", e, msg.data);
    }
  };
  es.onerror = () => es.close();
  return () => es.close();
}
